cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -1131,150 +1227,212 @@ aggregate_reader_metadata::select_row_groups(
+
   if (filter.has_value()) {
-    // Span of input row group indices for predicate pushdown
-    host_span<std::vector<size_type> const> input_row_group_indices;
-    std::vector<std::vector<size_type>> all_row_group_indices;
-    if (row_group_indices.empty()) {
-      std::transform(per_file_metadata.cbegin(),
-                     per_file_metadata.cend(),
-                     std::back_inserter(all_row_group_indices),
-                     [](auto const& file_meta) {
-                       std::vector<size_type> rg_idx(file_meta.row_groups.size());
-                       std::iota(rg_idx.begin(), rg_idx.end(), 0);
-                       return rg_idx;
-                     });
-      input_row_group_indices = host_span<std::vector<size_type> const>(all_row_group_indices);
-    } else {
-      input_row_group_indices = row_group_indices;
-    }
-    // Predicate pushdown: Filter row groups using stats and bloom filters
     std::tie(filtered_row_group_indices, num_row_groups_after_filters) =
       filter_row_groups(sources,
-                        input_row_group_indices,
-                        total_row_groups,
+                        current_row_group_indices,
+                        total_row_bounds_filtered_row_groups,
                         output_dtypes,
                         output_column_schemas,
                         filter.value(),
                         stream);
+
+    // If row groups were filtered, update the current span of row group indices as well as the
+    // relative rows to skip with respect to the first surviving row group
     if (filtered_row_group_indices.has_value()) {
-      row_group_indices =
+      // Update the current span of row group indices
+      current_row_group_indices =
         host_span<std::vector<size_type> const>(filtered_row_group_indices.value());
+
+      // Only need to update the relative rows to skip with respect to the first surviving row group
+      // if row bounds were previously applied
+      if (is_trimmed_row_groups) {
+        // Find the source index of the first non-empty row group
+        auto const first_non_empty_source_idx =
+          std::distance(current_row_group_indices.begin(),
+                        std::find_if(current_row_group_indices.begin(),
+                                     current_row_group_indices.end(),
+                                     [](auto const& indices) { return not indices.empty(); }));
+
+        // Update the relative rows to skip with respect to the first surviving row group
+        if (std::cmp_less(first_non_empty_source_idx, current_row_group_indices.size())) {
+          // Row group index of the first element in the first non-empty source
+          auto const first_non_empty_row_group_idx =
+            current_row_group_indices[first_non_empty_source_idx].front();
+
+          first_row_group_relative_rows_to_skip = std::max<int64_t>(
```

Comment:
Here we are resetting the `first_row_group_relative_rows_to_skip` as the difference between the absolute `rows_to_skip` and the row group offset of the first row group in the current span of row groups. We limit this to `zero` since it's possible the first row group starts at > `rows_to_skip` not needing any skip rows anymore.